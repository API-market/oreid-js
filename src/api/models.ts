import { AccountName, AccountType, ChainAccount, ChainNetwork } from '../common/models'

export enum RequestType {
  Get = 'get',
  Post = 'put',
}

export enum ApiEndpoint {
  AddPermission = 'account/add-permission',
  AppToken = 'app-token',
  CanAutoSign = 'transaction/can-auto-sign',
  ConvertOauthTokens = 'account/convert-oauth',
  CustodialMigrateAccount = 'custodial/migrate-account',
  CustodialNewAccount = 'custodial/new-user',
  DeleteTestUser = 'account/delete-test-user',
  CustodialSign = 'custodial/sign',
  GetUser = 'account/user',
  GetConfig = 'services/config',
  LoginUserWithToken = 'account/login-user-with-token',
  NewUserWithToken = 'account/new-user-with-token',
  PasswordLessSendCode = 'account/login-passwordless-send-code',
  PasswordLessVerifyCode = 'account/login-passwordless-verify-code',
  TransactionSign = 'transaction/sign',
}

/** Typical API Response (when no data returned) */
export type ApiMessageResult = {
  message?: string
  error?: string
  processId?: string
  success?: string
}

export type ApiResultWithErrorCode = {
  message?: string
  processId?: string
  errorCode?: string
  errorMessage?: string
}

// API params

export type CustodialMigrateAccountParams = {
  account: AccountName
  chainAccount: ChainAccount
  chainNetwork: ChainNetwork
  toType: AccountType
  userPassword: string
}

export type CustodialMigrateAccountApiBodyParams = {
  account: AccountName
  chain_account: ChainAccount
  chain_network: ChainNetwork
  to_type: AccountType
  user_password: string
}
