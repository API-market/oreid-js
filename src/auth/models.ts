import { AuthProvider, ChainAccount, ChainNetwork, ExternalWalletType } from '../common/models'

/** Raw data extracted from OAuth IDToken */
export type IdToken = {
  sub: string
  nickname: string
  phone_number: string
  email: string
  picture: string
  name: string
  email_verified?: boolean
}

/** Oauth JWTToken */
export type JWTToken = {
  iss: string
  sub: string
  aud: string
  exp: number
  nbf: number
  iat: number
  jti: string
} & { [key: string]: any }

export type LoginOptions = {
  provider?: AuthProvider
  idToken?: string
  chainAccount?: ChainAccount
  chainNetwork?: ChainNetwork
  code?: string
  email?: string
  phone?: string
  state?: string
  linkToAccount?: boolean
  returnAccessToken?: boolean
  returnIdToken?: boolean
}

export type LoginWithWalletOptions = {
  /** Optionally specify a specific account to select from wallet (some wallets don't support this) */
  chainAccount?: ChainAccount
  /** Optionally specify a specific blockchain to select from wallet (some wallets don't support this) */
  chainNetwork?: ChainNetwork
  walletType?: ExternalWalletType
}

export type LoginWithTokenOptions = {
  /** accessToken issued by provider specified (or ORE ID if no provider specified) */
  accessToken?: string
  idToken?: string
  /** the OAuth provider that issued the accessToken */
  provider?: AuthProvider
}

export type NewUserWithTokenOptions = {
  /** accessToken issued by provider specified */
  accessToken?: string
  idToken?: string
  /** the OAuth provider that issued the accessToken */
  provider?: AuthProvider
}
