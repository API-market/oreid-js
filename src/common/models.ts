export type ProcessId = string
export type AppAccessToken = string
export type AccountName = string
export type ChainAccount = string
export type PublicKey = string
export type PermissionName = string
export type Color = string

export enum ConfigType {
  Chains = 'chains',
}

/** Flavor of chain network */
export enum ChainPlatformType {
  algorand = 'algorand',
  eos = 'eos',
  ethereum = 'ethereum',
  ore = 'ore',
}

export enum ChainNetwork {
  AlgoMain = 'algo_main',
  AlgoBeta = 'algo_beta',
  AlgoTest = 'algo_test',
  EosMain = 'eos_main',
  EosKylin = 'eos_kylin',
  EosJugle = 'eos_jungle',
  EthMain = 'eth_main',
  EthRopsten = 'eth_ropsten',
  EthRinkeby = 'eth_rinkeby',
  OreMain = 'ore_main',
  OreTest = 'ore_test',
  DappKylin1 = 'kylin-dsp-1.liquidapps.io',
  DappKylin2 = 'kylin-dsp-2.liquidapps.io',
  TelosMain = 'telos_main',
  TelosTest = 'telos_test',
  WaxMain = 'wax_main',
  WaxTest = 'wax_test',
  UxMain = 'ux_main',
  PolygonMain = 'polygon_main',
  PolygonMumbai = 'polygon_mumbai',
}

/** Physical structure of account */
export enum AccountType {
  Native = 'native',
  Pending = 'pending',
  VirtualLiquid = 'liquid',
  VirtualNested = 'nested',
}

export enum LoginProvider {
  Custodial = 'custodial',
  Apple = 'apple',
  Email = 'email',
  Facebook = 'facebook',
  Github = 'github',
  Google = 'google',
  Instagram = 'instagram',
  Kakao = 'kakao',
  Line = 'line',
  LinkedIn = 'linkedin',
  Phone = 'phone',
  Twitch = 'twitch',
  Twitter = 'twitter',
}

export enum ExternalWalletType {
  AlgoSigner = 'algosigner',
  Keycat = 'keycat',
  Ledger = 'ledger',
  Lynx = 'lynx',
  Meetone = 'meetone',
  Metro = 'metro',
  Portis = 'portis',
  Scatter = 'scatter',
  SimpleEos = 'simpleos',
  TokenPocket = 'tokenpocket',
  WalletConnect = 'walletconnect',
  Web3 = 'web3',
  WhaleVault = 'whalevault',
}

/** All auth providers - including OreId login types (eg 'Facebook') and wallets (eg 'Scatter') */
export enum AuthProvider {
  // ORE ID LoginProvider
  Custodial = 'custodial',
  Apple = 'apple',
  Email = 'email',
  Facebook = 'facebook',
  Github = 'github',
  Google = 'google',
  Instagram = 'instagram',
  Kakao = 'kakao',
  Line = 'line',
  LinkedIn = 'linkedin',
  Phone = 'phone',
  Twitch = 'twitch',
  Twitter = 'twitter',
  // ExternalWalletType
  AlgoSigner = 'algosigner',
  Keycat = 'keycat',
  Ledger = 'ledger',
  Lynx = 'lynx',
  Meetone = 'meetone',
  Metro = 'metro',
  Portis = 'portis',
  Scatter = 'scatter',
  SimpleEos = 'simpleos',
  TokenPocket = 'tokenpocket',
  WalletConnect = 'walletconnect',
  Web3 = 'web3',
  WhaleVault = 'whalevault',
  // placeholder for signing - meaning non an external wallet
  OreId = 'oreid',
}

export enum ExternalWalletInterface {
  Transit = 'transit',
}

export enum ApiKeyUsedFor {
  AccountMigration = 'accountMigration',
  Airdrop = 'airdrop',
  AdminAccess = 'adminAccess',
  AutoSigning = 'autoSigning',
  ChangePassword = 'changePassword',
  CreateUser = 'createUser',
  ProxySigning = 'proxySigning',
  TokenFunding = 'tokenFunding',
}

/** type returned by oreid sevice config endpoint for chains */
export type SettingChainNetwork = {
  blockExplorerAccountUrl: string
  blockExplorerTxUrl: string
  chainCommunicationSettings: any
  createBridgeContract: string
  defaultTransactionSettings?: any
  /** Url of monitor */
  dfuseNetwork?: string
  hosts: SettingChainNetworkHost[]
  isTestNetwork: boolean
  monitorConfig?: {
    dfuseSupported: boolean
    endpoint: string
  }
  name: string
  logoUrl: string
  network: ChainNetwork
  type: ChainPlatformType
}

export type SettingChainNetworkHost = {
  chainId: string
  /** fork name used to designate a major version */
  forkName?: string
  host: string
  port: number
  protocol: 'https' | 'http'
}

// Chain specific

export type AlgorandMultiSigOptions = {
  version: number
  threshold: number
  addrs: string[]
}

/** helper type to index a JSON object */
export interface Lookup {
  [key: string]: any
}

export type JSONArray = Array<JSONValue>

export type JSONValue = string | number | boolean | Function | JSONObject | JSONArray

export interface JSONObject {
  [x: string]: JSONValue
}
