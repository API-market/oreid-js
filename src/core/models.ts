import { Request, Response } from 'express' // NOTE: We are only using types here - Express library won't be emitted in the build here
import {
  AccountName,
  AccountType,
  AlgorandMultiSigOptions, // TODO": should this be somewhere else?
  AppAccessToken,
  AuthProvider,
  ChainAccount,
  ChainNetwork,
  Color,
  ProcessId,
  PublicKey,
} from '../common/models'
import { TransitWalletProviderFactory } from '../transit'
import { UserInfo } from '../user/models'

export type OreIdOptions = {
  appId: string
  /** appKey is required to call the oreid API */
  apiKey?: string
  appName: string
  accessToken?: string
  idToken?: string
  authCallbackUrl?: string
  newAccountCallbackUrl?: string
  signCallbackUrl?: string
  backgroundColor?: Color
  /** whether you are using a proxy server - required for api calls or auth calls without idToken */
  isUsingProxyServer?: boolean
  oreIdUrl?: string
  serviceKey?: string
  setBusyCallback?: (isBusy: boolean) => void
  eosTransitWalletProviders?: TransitWalletProviderFactory[]
}

export enum RecoverAccountAction {
  Republic = 'republic',
}

// oreid-js

/** For rediecting to new-account endpoint - to create a chainAccount in an existing wallet */
export type NewAccountOptions = {
  account: AccountName // existing wallet account name (ore account)
  accountType: AccountType
  chainNetwork?: ChainNetwork
  accountOptions?: CreateOnChainAccountsOptions
  provider: AuthProvider
  state?: string
  processId?: ProcessId
  accessToken?: string
  idToken?: string
}

export type LoginOptions = {
  provider?: AuthProvider
  idToken?: string
  chainAccount?: ChainAccount
  chainNetwork?: ChainNetwork
  code?: string
  email?: string
  phone?: string
  state?: string
  linkToAccount?: boolean
  processId?: ProcessId
  returnAccessToken?: boolean
  returnIdToken?: boolean
}

export type DiscoverOptions = {
  provider: AuthProvider
  chainNetwork?: ChainNetwork
  oreAccount?: ChainAccount
  discoveryPathIndexList?: number[]
}

export type SignOptions = {
  account: AccountName
  allowChainAccountSelection?: boolean
  broadcast?: boolean
  callbackUrl?: string
  chainAccount?: ChainAccount // chainAccount not required for Ethereum - it can be defined by the account that signed the transaction
  chainNetwork: ChainNetwork
  expireSeconds?: number
  generateAutoSignCredential?: boolean
  /** Comma seperated string of accounts - for which OREID should add signatures */
  multiSigChainAccounts?: string
  preventAutoSign?: boolean
  processId?: ProcessId
  provider: AuthProvider
  returnSignedTransaction?: boolean
  signedTransaction?: string
  signExternalWithOreId?: boolean
  state?: string
  transaction?: string
  transactionChainAccount?: string
  transactionRecordId?: string
  userPassword?: string
  userPasswordEncrypted?: string
}

export type AuthResponse = {
  account: AccountName
  accessToken?: string
  idToken?: string
  errors?: string[]
  processId?: ProcessId
  state?: string
}

export type NewAccountResponse = {
  chainAccount?: string
  errors?: string[]
  processId?: ProcessId
  state?: string
}

export type SignResponse = {
  signedTransaction?: string
  transactionId: string
  errors?: string[]
  processId?: ProcessId
  state?: string
}

// Todo: implement passwordReset flow/url
export type passwordResetOptions = {
  provider: AuthProvider
  chainAccount?: ChainAccount
  chainNetwork?: ChainNetwork
  state?: string
  currentAccountPassword?: string // if the user's current password is known, it can be passed in to appAccessToken request
  processId?: ProcessId
}

export type GetRecoverAccountUrlResult = string

export type LoginWithOreIdResult = {
  accessToken?: string
  loginUrl?: string
  errors?: string
  processId?: ProcessId
}

export type NewAccountWithOreIdResult = {
  newAccountUrl: string
  errors?: string
}

export type SignWithOreIdResult = {
  processId?: ProcessId
  signedTransaction?: string
  transactionId?: string
  signUrl?: string
  errors?: string
}

export type SignStringParams = {
  account: AccountName
  provider: AuthProvider
  chainAccount?: ChainAccount
  chainNetwork: ChainNetwork
  string: string
  message: string
}

// Function Params

export type GetOreIdNewAccountUrlParams = NewAccountOptions & {
  callbackUrl: string
  backgroundColor?: string
}

export type GetOreIdAuthUrlParams = LoginOptions & {
  callbackUrl: string
  backgroundColor?: string
}

export type GetOreIdRecoverAccountUrlParams = LoginOptions & {
  account: AccountName
  backgroundColor?: string
  callbackUrl: string
  recoverAction: RecoverAccountAction
  overrideAppAccessToken?: AppAccessToken
}

export type AppAccessTokenMetadata = {
  paramsNewAccount?: NewAccountAppTokenParams
  newAccountPassword?: string
  currentAccountPassword?: string
  secrets?: {
    type: string
    value: string
  }[]
}

/** params for calling new-account service web endpoint (sent via AppAccessToken request) */
export type NewAccountAppTokenParams = {
  account: AccountName // wallet account (ore account name)
  accountType: AccountType
  chainNetwork?: ChainNetwork // chainNetwork to create new account on
  accountOptions: CreateOnChainAccountsOptions
}

export type CreateOnChainAccountsOptions = {
  keys?: {
    publicKeys: {
      owner?: PublicKey
      active: PublicKey
    }
  }
  multisigOptions?: AlgorandMultiSigOptions // || future multisig option types
}

type ParamsForRequest = {
  appId?: string
  accessToken?: string
  chainAccount?: string
  idToken?: string
  processId?: ProcessId
  state?: string
  signedTransaction?: string
  transactionId?: string
  user?: UserInfo
}

type ParamsForResponse = {
  myField?: string
}

export type RequestWithParams = Request & ParamsForRequest
export type ResponseWithParams = Response & ParamsForResponse
