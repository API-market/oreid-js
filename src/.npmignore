apimarket_config.json
package-lock.json
dist
node_modules
