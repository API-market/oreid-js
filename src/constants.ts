export const transitProviderAttributes = {
  ledger: {
    providerId: 'ledger',
    requiresLogin: true,
    supportsDiscovery: true,
    supportsSignArbitrary: false,
    requiresLogoutLoginToDiscover: false,
    defaultDiscoveryPathIndexList: [0, 1, 2],
    helpText: {
      login: 'This wallet doesn’t require you to login',
      sign: 'Connect and unlock your Ledger with your PIN code. Launch the EOS app on the device.',
      discover: 'Connect and unlock your Ledger with your PIN code. Launch the EOS app on the device.',
      versionsRequired: 'You need to have recent versions of your browser, Ledger firmware, and the Ledger EOS app. Click here for more details.'
    }
  },
  lynx: {
    providerId: 'EOS Lynx',
    requiresLogin: true,
    supportsDiscovery: false,
    supportsSignArbitrary: true,
    requiresLogoutLoginToDiscover: true,
    helpText: {
      login: 'EOS Lynx requires logging in through the EOS Lynx app.',
      sign: '',
      discover: '',
      versionsRequired: ''
    }
  },
  meetone: {
    providerId: 'meetone_provider',
    requiresLogin: false,
    supportsDiscovery: false,
    supportsSignArbitrary: true,
    requiresLogoutLoginToDiscover: true,
    helpText: {
      login: 'This wallet doesn’t require you to login',
      sign: '',
      discover: '',
      versionsRequired: ''
    }
  },
  metro: {
    providerId: 'metro',
    requiresLogin: false,
    supportsDiscovery: false,
    supportsSignArbitrary: false,
    requiresLogoutLoginToDiscover: false,
    helpText: {
      login: 'This wallet doesn’t require you to login',
      sign: '',
      discover: '',
      versionsRequired: ''
    }
  },
  scatter: {
    providerId: 'scatter',
    requiresLogin: true,
    supportsDiscovery: false,
    supportsSignArbitrary: true,
    requiresLogoutLoginToDiscover: true,
    helpText: {
      login: 'Scatter requires logging in through the Scatter app.',
      sign: '',
      discover: '',
      versionsRequired: ''
    }
  },
  tokenpocket: {
    providerId: 'TokenPocket',
    requiresLogin: true,
    supportsDiscovery: false,
    supportsSignArbitrary: true,
    requiresLogoutLoginToDiscover: true,
    helpText: {
      login: 'This wallet doesn’t require you to login',
      sign: '',
      discover: '',
      versionsRequired: ''
    }
  },
  portis: {
    providerId: 'PortisProvider',
    requiresLogin: true,
    supportsDiscovery: false,
    supportsSignArbitrary: false,
    requiresLogoutLoginToDiscover: true,
    helpText: {
      login: 'Portis requires logging in through the Portis app.',
      sign: '',
      discover: '',
      versionsRequired: ''
    }
  },
  whalevault: {
    providerId: 'whalevault',
    requiresLogin: true,
    supportsDiscovery: false,
    supportsSignArbitrary: true,
    requiresLogoutLoginToDiscover: true,
    helpText: {
      login: 'Whalevault requires logging in through the Whalevault app.',
      sign: '',
      discover: '',
      versionsRequired: ''
    }
  },
  simpleos: {
    providerId: 'simpleos',
    requiresLogin: true,
    supportsDiscovery: false,
    supportsSignArbitrary: false,
    requiresLogoutLoginToDiscover: true,
    helpText: {
      login: 'Simpleos requires logging in through the Simpleos app.',
      sign: '',
      discover: '',
      versionsRequired: ''
    }
  },
  keycat: {
    providerId: 'Keycat',
    requiresLogin: true,
    supportsDiscovery: false,
    supportsSignArbitrary: true,
    requiresLogoutLoginToDiscover: true,
    helpText: {
      login: 'Keycat requires logging in through the Keycat app.',
      sign: '',
      discover: '',
      versionsRequired: ''
    }
  }
};

export const ualProviderAttributes = {
  scatter: {
    requiresLogin: true,
    supportsSignArbitrary: true,
    helpText: {
      login: 'Scatter requires logging in through the Scatter app.',
      sign: '',
      discover: '',
      versionsRequired: ''
    }
  },
  ledger: {
    requiresLogin: true,
    supportsSignArbitrary: false,
    helpText: {
      login: 'This wallet doesn’t require you to login',
      sign: 'Connect and unlock your Ledger with your PIN code. Launch the EOS app on the device.',
      discover: 'Connect and unlock your Ledger with your PIN code. Launch the EOS app on the device.',
      versionsRequired: 'You need to have recent versions of your browser, Ledger firmware, and the Ledger EOS app. Click here for more details.'
    }
  },
  lynx: {
    requiresLogin: false,
    supportsSignArbitrary: true,
    helpText: {
      login: 'This wallet doesn’t require you to login',
      sign: '',
      discover: '',
      versionsRequired: ''
    }
  },
  meetone: {
    requiresLogin: false,
    supportsSignArbitrary: true,
    helpText: {
      login: 'This wallet doesn’t require you to login',
      sign: '',
      discover: '',
      versionsRequired: ''
    }
  },
  tokenpocket: {
    requiresLogin: false,
    supportsSignArbitrary: true,
    helpText: {
      login: 'This wallet doesn’t require you to login',
      sign: '',
      discover: '',
      versionsRequired: ''
    }
  }
};

export const providersNotImplemented = [
  'metro'
];

export const supportedTransitProviders = Object.keys(transitProviderAttributes);
export const supportedUALProviders = Object.keys(ualProviderAttributes);
