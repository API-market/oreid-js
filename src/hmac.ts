import CryptoJS from 'crypto-js'
import Axios from 'axios'
import Helpers from './helpers'

export function generateHmac(secret: string, data: string) {
  const hmac = CryptoJS.HmacSHA256(data, secret)
  return hmac.toString()
}

/** Generate HMAC for url string and append it to end e.g. http:/nnnnnn&hmac=xxx */
export async function appendHmacToUrl(useProxyServer: boolean, apiKey: string, url: string) {
  if (!useProxyServer && Helpers.isNullOrEmpty(apiKey)) {
    throw new Error('Cant generate hmac. apiKey missing or empty - we need it if we arent using a proxy server')
  }
  if (Helpers.isNullOrEmpty(url)) {
    throw new Error('Cant generate hmac. url missing or empty')
  }

  let hmac: string
  if (useProxyServer) {
    try {
      // get an hmac from the proxy server endpoint (since it has the secret apiKey)
      const response = await Axios.post('/oreid/hmac', JSON.stringify({ data: url }), {
        headers: { 'Content-Type': 'application/json' },
      })
      hmac = response?.data?.hmac
    } catch (networkError) {
      const error = this.getErrorFromAxiosError(networkError)
      throw error
    }
  } else {
    hmac = generateHmac(apiKey, url)
  }

  const urlEncodedHmac = encodeURIComponent(hmac)
  // correct hmac already in place
  if (url.includes(`&hmac=${urlEncodedHmac}`)) {
    return url
  }
  return `${url}&hmac=${urlEncodedHmac}`
}
