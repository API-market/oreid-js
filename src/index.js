const { isNullOrEmpty, jwtDecodeSafe, log, tokenHasExpired, tryParseJSON, urlParamsToArray } = require("./helpers.js");
// const StorageHandler = require("./storage.js");
const Base64 = require('js-base64').Base64;
const fetch = require('node-fetch');
import StorageHandler from './storage';

const APPID_CLAIM_URI = "https://oreid.aikon.com/appId";

class OreId {

    constructor(options) {
        this.options;
        this.appAccessToken;
        this.user;
        this.storage = new StorageHandler();
        this.validateOptions(options);
    }

    /*
        Validates startup options
    */
    validateOptions(options) {
        let { appId, apiKey, oreIdUrl } = options;
        let errorMessage = ''

        if (!appId) {
            errorMessage += `\n --> Missing required parameter - appId. You can get an appId when you register your app with ORE ID.`
        }
        if (!apiKey) {
            errorMessage += `\n --> Missing required parameter - apiKey. You can get an apiKey when you register your app with ORE ID.`
        }
        if (!oreIdUrl) {
            errorMessage += `\n --> Missing required parameter - oreIdUrl. Refer to the docs to get this value.`
        }
        if (errorMessage != '') {
            throw new Error(`Options are missing or invalid. ${errorMessage}`)
        }

        this.options = options;
    };

    /*
        load user from local storage and call api to get latest info
    */
    async getUser(account) {
        if (account) {
            const user = await this.getUserInfoFromApi(account); //get the latest user data
            return user;
        }
        // Check in state
        if(this.user) {
            return this.user;
        }
        // Check local storage
        let result = this.loadUserLocally();
        return result;
    }

    /*
        Checks for the expiration of the locally cached app-access-token
        Refreshes token if needed using getNewAppAccessTokenFromApi()
    */
    async getAccessToken() {
        //check for expiration and renew token if expired
        if(!this.appAccessToken || tokenHasExpired(this.appAccessToken)) {
            await this.getNewAppAccessToken(); //call api
        }
        return this.appAccessToken;
    };

    /*
        Returns a fully formed url to call the auth endpoint
    */
    async getOreIdAuthUrl(authOptions) {
        let { loginType, callbackUrl, backgroundColor, state } = authOptions;
        let { oreIdUrl } = this.options;

        if(!loginType || !callbackUrl) {
            throw new Error(`Missing a required parameter`);
        }

        let appAccessToken = await this.getAccessToken();
        let encodedStateParam = (state) ? `&state=${Base64.encode(JSON.stringify(state))}` : "";

        return `${oreIdUrl}/auth#app_access_token=${appAccessToken}?provider=${loginType}?callback_url=${encodeURIComponent(callbackUrl)}?background_color=${backgroundColor}${encodedStateParam}`;
    };

    /*
        Returns a fully formed url to call the sign endpoint
    */
    async getOreIdSignUrl(signOptions) {
        let { account, broadcast, callbackUrl, chain, state, transaction } = signOptions;
        let { oreIdUrl } = this.options;

        if(!transaction || !account || !callbackUrl || !chain) {
            throw new Error(`Missing a required parameter`);
        }

        let appAccessToken = await this.getAccessToken();
        let encodedTransaction = Base64.encode(JSON.stringify(transaction));
        let encodedStateParam = (state) ? `&state=${Base64.encode(JSON.stringify(state))}` : "";

        return `${oreIdUrl}/sign#app_access_token=${appAccessToken}&account=${account}&callback_url=${encodeURIComponent(callbackUrl)}&chain=${chain}&broadcast=${broadcast}&transaction=${encodedTransaction}${encodedStateParam}`;
    };

    /*
        Extracts the response parameters on the /auth callback URL string
    */
    handleAuthResponse(callbackUrlString) {
        //Parses error codes and returns an errors array
        //(if there is an error_code param sent back - can have more than one error code - seperated by a ‘&’ delimeter
        let params = urlParamsToArray(callbackUrlString);
        let state;
        let { account, state:encodedState } = params;
        let errors = this.getErrorCodesFromParams(params);

        if(encodedState) {
            //Decode base64 parameters
            state = JSON.parse(Base64.decode(encodedState));
        }
        return {account, state, errors};
    };

    /*
        Extracts the response parameters on the /sign callback URL string
    */
    handleSignResponse(callbackUrlString) {
        let signedTransaction;
        let state;
        let params = urlParamsToArray(callbackUrlString);
        let {signed_transaction:encodedTransaction, state:encodedState} = params;
        let errors = this.getErrorCodesFromParams(params);

        if(!errors) {
            //Decode base64 parameters
            signedTransaction = tryParseJSON(Base64.decode(encodedTransaction));
            state = JSON.parse(Base64.decode(encodedState));
        }
        return {signedTransaction, state, errors};
    };

    /*
        Calls the {oreIDUrl}/api/app-token endpoint to get the appAccessToken
    */
    async getNewAppAccessToken() {
        let responseJson = await this.callOreIdApi(`app-token`)
        let { appAccessToken } = responseJson;
        this.appAccessToken = appAccessToken;
        let decodedToken = jwtDecodeSafe(appAccessToken);
        // this.appId = decodedToken[APPID_CLAIM_URI]; //Get the appId from the app token
    };

    /*
        Get the user info from ORE ID for the given user account
    */
    async getUserInfoFromApi(account) {
        let responseJson = await this.callOreIdApi(`account/user?account=${account}`)
        let userInfo = responseJson;
        this.saveUserLocally(userInfo);
        await this.loadUserLocally(); //ensures this.user state is set
        return userInfo;
    };

    /*
        Adds a public key to an account with a specific permission name 
        The permission name must be one defined in the App Registration record (Which defines its parent permission as well as preventing adding rougue permissions)
        This feature allows your app to hold private keys locally (for certain actions enabled by the permission) while having the associated public key in the user's account
        chainAccount is the name of the account on the chain - 12/13-digit string on EOS and Ethereum Address on ETH
    */
   async addPermission(account, chain, chainAccount, publicKey, permission) {
        let responseJson = await this.callOreIdApi(`account/add-permission?account=${account}&chain=${chain}&chain-account=${chainAccount}&public-key=${publicKey}&permission=${permission}`)
        //if failed, error will be thrown
    };

    /*
        Get the user info from ORE ID for the given user account
    */
    async getUserWalletInfo(account) {
        throw Error('Not Implemented');
        // let responseJson = await this.callOreIdApi(`wallet?account=${account}`)
        // let userWalletInfo = responseJson;
        // return {userWalletInfo, errors};
    };

    /*
        Helper function to call api endpoint and inject api-key
    */
    async callOreIdApi(endpointAndParams) {
        let { apiKey, oreIdUrl } = this.options;
        let url = `${oreIdUrl}/api/${endpointAndParams}`;
        let response = await fetch(url, {
            headers: { 'api-key' : apiKey }}
        );

        let responseJson = await response.json();
        let { error, message } = responseJson;
        if (error) {
            throw new Error(error);
        };
        return responseJson;
    }

    /*
        Params is a javascript object representing the parameters parsed from an URL string
    */
    getErrorCodesFromParams(params) {
        let errorCodes;
        let errorString = params["error_code"];
        let errorMessage = params["error_message"];
        if(errorString) {
            errorCodes = errorString.split(/[/?/$&]/);
        }
        if(errorCodes || errorMessage) {
            errorCodes = errorCodes || [];
            errorCodes.push(errorMessage);
        };
        return errorCodes;
    };

    /*
        We don't really maintain a logged-in state 
        However, we do have local cached user data, so clear that
    */
    logout() {
        //clear local state
        this.clearLocalState();
    }

    /*
        Local state
    */

    userKey() {
        return `oreid.${this.options.appId}.user`;
    }

    saveUserLocally(user) {
        if(isNullOrEmpty(user)) { return; }
        this.user = user;
        let serialized = JSON.stringify(this.user);
        this.storage.setItem(this.userKey(), serialized);
    }

    loadUserLocally() {
        let serialized = this.storage.getItem(this.userKey());
        //user state does not exist
        if(isNullOrEmpty(serialized)) {
            this.user = null;
            return null;
        }
        this.user = JSON.parse(serialized);
        return this.user;
    }

    async clearLocalState() {
        this.storage.removeItem(this.userKey());
    }

}

module.exports = {
    OreId
}
