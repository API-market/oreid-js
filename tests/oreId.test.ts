/* eslint-disable jest/no-mocks-import */
import OreId from '../src/oreId'
import demoChainNetworks from '../src/testHelpers/__mocks__/chainNetworks.json'
import { OreIdOptions, AuthProvider } from '../src/models'
import { generateHmac } from '../src/hmac'
import { defaultOreIdServiceUrl } from '../src/constants'

describe('OreId', () => {
  let oreId: OreId
  let options: OreIdOptions = {
    appName: 'testrunner',
    appId: 'demo_0097ed83e0a54e679ca46d082ee0e33a',
    apiKey: 'demo_k_97b33a2f8c984fb5b119567ca19e4a49',
    oreIdUrl: defaultOreIdServiceUrl,
  }

  beforeEach(() => {
    oreId = new OreId(options)
  })

  describe('Errors without the required params', () => {
    it('Throws the correct message with no params', () => {
      const error = `Options are missing or invalid. 
 --> Missing required parameter - appId. You can get an appId when you register your app with ORE ID.`
      expect(() => {
        oreId = new OreId(null)
      }).toThrow(Error(error))
    })

    it('Throws an error without an appId', () => {
      const appIdError = `Options are missing or invalid. 
 --> Missing required parameter - appId. You can get an appId when you register your app with ORE ID.
 --> You cant include the apiKey (or serviceKey) when creating an instance of OreId that runs in the browser. This is to prevent your keys from being visible in the browser. If this app runs solely in the browser (like a Create React App), you need to set-up a proxy server to protect your keys. Refer to https://github.com/TeamAikon/ore-id-docs. Note: You wont get this error when using the appId and apiKey for a demo app (appId starts with demo_).`
      expect(() => {
        oreId = new OreId({ ...options, appId: '' })
      }).toThrowError(Error(appIdError))
    })
  })

  it('Doesnt error when it has the required values', () => {
    expect(oreId).toBeTruthy()
  })

  it('sets the chainNetworks correctly when initializing', async () => {
    await oreId.chainNetworks()
    expect(oreId.cachedChainNetworks).toEqual(demoChainNetworks)
  })

  describe('Login', () => {
    options = {
      ...options,
      authCallbackUrl: 'http://localhost.com',
      backgroundColor: '',
    }

    const loginOptions = {
      code: '12345',
      email: 'test@test.com',
      phone: '+1555555555',
      provider: AuthProvider.Google,
      state: 'abc',
    }

    beforeEach(() => {
      oreId = new OreId(options)
    })

    it('errors without `provider` and `callbackUrl`', async () => {
      try {
        await oreId.login({
          ...loginOptions,
          provider: null,
        })
      } catch (error) {
        expect(error.message).toBe('Missing a required parameter')
      }
    })

    it('Throws an error if the provider is not provided', async () => {
      try {
        await oreId.login({
          ...loginOptions,
          provider: null,
        })
      } catch (error) {
        expect(error.message).toBe('Missing a required parameter')
      }
    })

    it('logs in with oreid', async () => {
      const result = await oreId.login(loginOptions)
      expect(result).toEqual({
        errors: null,
        loginUrl:
          'https://service.oreid.io/auth#provider=google&code=12345&email=test%40test.com&phone=%2B1555555555&callback_url=http%3A%2F%2Flocalhost.com&background_color=&state=abc&app_id=demo_0097ed83e0a54e679ca46d082ee0e33a&hmac=j1faPFT4zDNMssG4Gh13qDpz8n2QcQqBD0SuAGeoyy4%3D',
      })
    })

    it('Creates an HMAC', () => {
      const loginUrl =
        'https://service.oreid.io/auth#provider=google&code=12345&email=test@test.com&phone=%2B1555555555&callback_url=http%3A%2F%2Flocalhost.com&background_color=&state=abc&app_access_token=12345667'
      const hmac = generateHmac('demo_k_97b33a2f8c984fb5b119567ca19e4a49', loginUrl)
      expect(hmac).toEqual('LIuhxTG3o6ZZPaqD2CA+NIZ9oi5Lg11lKJNPUsN+Yto=')
    })

    // describe('Logins with transit', () => {
    //   const fakeScatterProvider = {
    //     id: 'scatter'
    //   };

    //   options = {
    //     ...options,
    //     eosTransitWalletProviders: [fakeScatterProvider]
    //   };

    //   beforeEach(() => {
    //     oreId = new OreId(options);
    //   });

    //   it('tries to login with scatter', async () => {
    //     const result = await oreId.login({ ...loginOptions, provider: 'scatter' });
    //     expect(result).toEqual('');
    //   });
    // });
  })
})
