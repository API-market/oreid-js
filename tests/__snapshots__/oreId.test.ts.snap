// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OreId sets the chainNetworks correctly when initializing 1`] = `undefined`;
